scheduler/scheduler-server/src/main/java/org/ow2/proactive/scheduler/core/db/SchedulerDBManager.java

```diff
@@ -2039,6 +2044,34 @@ public void newJobSubmitted(final InternalJob job) {
         });
     }
 
+    private void replaceSystemVariables(JobData jobData) {
+        Map<String, Serializable> variableReplacement = new LinkedHashMap<>();
+        variableReplacement.put(SchedulerVars.PA_JOB_ID.name(), jobData.getId());
+        variableReplacement.put(SchedulerVars.PA_JOB_NAME.name(), jobData.getJobName());
+        variableReplacement.put(SchedulerVars.PA_USER.name(), jobData.getOwner());
+
+        jobData.getVariables().values().forEach(jobVariable ->
+                variableReplacement.put(jobVariable.getName(), jobVariable.getValue()));
+        jobData.getVariables().values().forEach(runtimeVariable -> {
+               String originalValue = runtimeVariable.getValue();
+               runtimeVariable.setValue(VariableSubstitutor.filterAndUpdate(originalValue, variableReplacement));
+            });
+    }
+
+    private void replaceSystemVariables(InternalJob internalJob) {
+        Map<String, Serializable> variableReplacement = new LinkedHashMap<>();
+        variableReplacement.put(SchedulerVars.PA_JOB_ID.name(), internalJob.getId());
+        variableReplacement.put(SchedulerVars.PA_JOB_NAME.name(), internalJob.getName());
+        variableReplacement.put(SchedulerVars.PA_USER.name(), internalJob.getOwner());
+
+        internalJob.getVariables().values().forEach(jobVariable ->
+                variableReplacement.put(jobVariable.getName(), jobVariable.getValue()));
+        internalJob.getVariables().values().forEach(runtimeVariable -> {
+            String originalValue = runtimeVariable.getValue();
+            runtimeVariable.setValue(VariableSubstitutor.filterAndUpdate(originalValue, variableReplacement));
+        });
```

Comment:
I think you can do both processes in a single `forEach()` to avoid streaming twice